### PR TITLE
Windows resume unlock detect

### PR DESCRIPTION
**Pull request by mary-dcouto**
This PR contains:
endpoints change function call when windows power event or unlock event is detected.
Update tun ip command fix
Enable/Disable ziti tun 

(I merged main branch into this branch, so you will see some irrelevant changes too)

### CODE REVIEW

File: lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c [hunk 1]
```diff
@@ -157,7 +157,7 @@ static void ziti_dump_to_file(void *ctx, char* outputFile) {
 static void disconnect_identity(ziti_context ziti_ctx, void *tnlr_ctx) {
     ZITI_LOG(INFO, "Disconnecting Identity %s", ziti_get_identity(ziti_ctx)->name);
     remove_intercepts(ziti_ctx, tnlr_ctx);
-    ziti_shutdown(ziti_ctx);
+    // ziti_shutdown(ziti_ctx); // causes the crash
```
Comment (scareything) [inline lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c:160]: Should this comment mention issue 275?

File: lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c [hunk 2]
```diff
@@ -519,8 +519,31 @@ static int process_cmd(const tunnel_command *cmd, command_cb cb, void *ctx) {
...
+                ziti_endpoint_state_change(inst->ztx, tunnel_status_change_cmd.woken, tunnel_status_change_cmd.unlocked);
+                ZITI_LOG(DEBUG, "Endpoint status change function is invoked for %s with woken %d and unlocked %d", inst->identifier, tunnel_status_change_cmd.woken, tunnel_status_change_cmd.unlocked);
+            }
+            result.success = true;
+            free_tunnel_status_change(&tunnel_status_change_cmd);
             break;
+        }
+
+        case TunnelCommand_Unknown:
+        break;
```
Comment (scareything) [inline lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c:546]: indentation

File: programs/ziti-edge-tunnel/netif_driver/windows/tun.c [hunk 3]
```diff
@@ -214,10 +222,6 @@ netif_driver tun_open(struct uv_loop_s *loop, uint32_t tun_ip, uint32_t dns_ip,
     uv_unref((uv_handle_t *) &tun->route_timer);
     uv_timer_start(&tun->route_timer, refresh_routes, ROUTE_REFRESH, ROUTE_REFRESH);
 
-    if (dns_ip) {
```
Comment (scareything) [inline programs/ziti-edge-tunnel/netif_driver/windows/tun.c:217]: `dns_ip` is now an unused argument here. it should be removed from the argilst for the windows implementation of `tun_open`

File: programs/ziti-edge-tunnel/ziti-edge-tunnel.c [hunk 4]
```diff
@@ -1129,6 +1130,19 @@ static void on_event(const base_event *ev) {
...
 
+static char* normalize_host(char* hostname) {
+    size_t len = strlen(hostname);
+    char* hostname_new = calloc(len+1, sizeof(char));
+    // add . in the beginning of the hostname
+    if (hostname[len-1] == ".") {
+        // remove the . from the end of the hostname
+        snprintf(hostname_new, len * sizeof(char), ".%s", hostname);
+    } else {
+        sprintf(hostname_new,".%s", hostname);
```
Comment (scareything) [inline programs/ziti-edge-tunnel/ziti-edge-tunnel.c:1141]: This creates an invalid memory access at the end of `hostname_new` where the terminating '\0' should be.

`hostname_new` has strlen(hostname)+1 bytes; the extra byte gives you a place for the terminating nut byte. If you also want to add a character to the beginning of the string then you need to allocate `strlen(hostname)+2`.

File: programs/ziti-edge-tunnel/ziti-edge-tunnel.c [hunk 5]
```diff
@@ -1129,6 +1130,19 @@ static void on_event(const base_event *ev) {
     }
 }
 
+static char* normalize_host(char* hostname) {
+    size_t len = strlen(hostname);
+    char* hostname_new = calloc(len+1, sizeof(char));
+    // add . in the beginning of the hostname
+    if (hostname[len-1] == ".") {
```
Comment (scareything) [inline programs/ziti-edge-tunnel/ziti-edge-tunnel.c:1137]: comparing `char` which is an integer type to a string (denoted by double quotes) is not valid, and will not work the way you are probably expecting. CLIon and probably any halfway decent IDE will warn you about this:
image

In this case you want to use single-quotes for your constant:
```
if (hostname[len-1] == '.') {
```

File: programs/ziti-edge-tunnel/ziti-edge-tunnel.c [hunk 6]
```diff
@@ -1137,6 +1151,35 @@ static int run_tunnel(uv_loop_t *ziti_loop, uint32_t tun_ip, uint32_t dns_ip, co
...
+    }
+    if (nrpt_effective) {
+        model_map *domains = get_connection_specific_domains();
+        const char *key;
+        bool status;
+        model_map *normalized_domains = calloc(1, sizeof(model_map));
+        model_map_iter it = model_map_iterator(domains);
+        while (it != NULL) {
+            char *key = model_map_it_key(it);
+            model_map_set(normalized_domains, normalize_host(key), true);
```
Comment (scareything) [inline programs/ziti-edge-tunnel/ziti-edge-tunnel.c:1168]: `true` is not a valid value for a pointer argument. Is the value of this map ever used? I think it isn't - it looks like the map is being used as a list of keys, so you could set the values to NULL.

File: programs/ziti-edge-tunnel/ziti-edge-tunnel.c [hunk 7]
```diff
@@ -1137,6 +1151,35 @@ static int run_tunnel(uv_loop_t *ziti_loop, uint32_t tun_ip, uint32_t dns_ip, co
...
+            it = model_map_it_remove(it);
+        }
+        model_map_clear(domains, (_free_f) free);
+        free(domains);
+        struct add_service_nrpt_req *add_svc_req_data = calloc(1, sizeof(struct add_service_nrpt_req));
+        add_svc_req_data->hostnames = normalized_domains;
+        add_svc_req_data->dns_ip = get_dns_ip();
+
+        uv_async_t *ar = calloc(1, sizeof(uv_async_t));
```
[the remark this answers is not in the document]
Comment (mary-dcouto) [reply programs/ziti-edge-tunnel/ziti-edge-tunnel.c:1177]: I thought the below statement in add_nrpt_rules(uv_async_t *ar) function defined in windows-script.c will free ar.

uv_close((uv_handle_t *) ar, (uv_close_cb) free);

**Comment by mary-dcouto** [comment]
I merged the main branch and the next PR which was approved into this PR branch. Fixed the code as per the comments. Please review again

**Review by scareything — APPROVED** [review]

**Comment by mary-dcouto** [comment]
I fixed the code as per the latest review comments. I am merging the code as you have approved it. Please let me know if you have any more comments on this PR

**Pull request merged**